tests: Bluetooth: Mesh: add unit tests for blob_io_flash module

KKopyscinski opened the pull request:
This PR adds unit test for blob_io_flash module and fixed bug with `chunk_wr` by saving stream mode.

KKopyscinski committed:
Bluetooth: Mesh: save stream mode in io_open()

Mode affects access to flash aread for `block_start`. When mode is set
to `BT_MESH_BLOB_READ` function return early, and because
`BT_MESH_BLOB_READ = 0` this is default behaviour. For write flash area
must be erased to allow driver to write there new data - bits can only
be pulled down, so overwrite will not be permitted.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

KKopyscinski committed:
tests: Bluetooth: Mesh: add unit tests for blob_io_flash module

This tests functions accessible from blob_io_flash via `bt_mesh_blob_io`
object. Whole partition of slot1 is filled with test data using
flash_map API and then read using blob_io_flash, by reading it as blocks
and chunks. `chunk_wr` is tested my writing sample data of size of slot1
partition and checked if each chunk is written correctly, and if whole
partition is filled with correct data at the end. Both of these test
use `bt_mesh_blob_io_flash_init` and `io_open` to prepare stream and
call `block_start` at every new block to check if flash is prepared
correctly.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

alxelax commented on tests/bluetooth/blob_io_flash/testcase.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  blob_io_flash.flash_io:
+    platform_allow: nrf51dk_nrf51422 qemu_x86 native_posix native_posix_64
```
nrf51 is a deprecated platform with low capabilities. No needs to add for the tests. If you need to run it on real hardware please use `nrf52840dk_nrf52840`.

Did you run this test on all listed platforms?

KKopyscinski replied:
I see now that there is a problem on qemu, I'll test the rest too

KKopyscinski replied:
I'm removing targets other than native, because I use arrays with test data of very big sizes to check BLOB data written and read from flash. Testing this on ARM platforms, including QEMU, could be done in the future as BSIM test.

KKopyscinski replied:
Then we can consider that data verification is covered here and module running in real BLOB transfer will happen on simulated HW

alxelax replied:
I do not think it is worth to have two tests (unit and bsim) for the same functionality. This is overcomplication. Did you try to run it on nrf52840?

KKopyscinski replied:
Because I'm using big arrays to verify what was written into/read from flash it's not possible to use it on any real HW or qemu, because of flash limitation. And I think it's necessary to assure that correct data is stored in flash - that's how I found bug in `block_start`.

One idea is to limit flash area size so much that it would fit on stack - that way I can write whole partition and read it out to compare with control data. I suspect this could be done with `.dts` overlay, but I'm not sure. I don't want to just use data smaller than partition: I want to stress test this, write many chunks with maximum size, reach end of flash, write more and see if error triggers. IIRC partition size is something around 43kB, no real board has RAM that big. But this test is not running on embedded, so using native_posix I can afford making arrays this big

alxelax replied:
I'm Ok to run it only on native_posix platform. This is the unit test. The platform really doesn't matter. I worry the only thing there shouldn't be platforms that do not work with test.

alxelax commented on tests/bluetooth/blob_io_flash/src/main.c:
```diff
@@ -0,0 +1,256 @@
+		/* `block_start` in write mode will erase flash pages that can fit block.
+		 * Assert that at least block size of data was not erased in read mode
+		 */
+		flash_area_read(fa, block.offset, ctrl_data, block.size);
```
```suggestion
		flash_area_read(blob_flash_stream.area, block.offset, ctrl_data, block.size);
```

alxelax commented:
There is build issue in BLOB client.

PavelVPV commented on tests/bluetooth/blob_io_flash/src/main.c:
```diff
@@ -0,0 +1,256 @@
+				     : remaining;
+		block.offset = block_idx * (1 << block_size_to_log(block.size));
+
+		/* BLOB Client should do nothing to flash area as it's in read mode */
```
I guess the comment is incorrect since it is write test.

KKopyscinski committed:
Bluetooth: Mesh: fix blob_cli.c build with Clang

There should be logic OR, not bitwise OR in update_missing_chunks.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

PavelVPV commented on tests/bluetooth/blob_io_flash/testcase.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  blob_io_flash.flash_io:
+    platform_allow: native_posix
+    tags: blob_io_flash
+    integration_platforms:
+      - native_posix
```
```suggestion
  bluetooth.mesh.blob_io_flash:
    platform_allow: native_posix
    tags: bluetooth mesh
    integration_platforms:
      - native_posix
```

PavelVPV reviewed (commented):
Looks very good. One suggestion. I don't think it is good to litter in `test/bluetooth` folder because we can start adding more unit tests. So I'd suggest to move the app located in `test/bluetooth/mesh` to `tests/bluetooth/mesh/basic` (I couldn't give any better name, welcome with suggestions) and then move the new test to `test/bluetooth/mesh/blob_io_flash`? Or I can do this in a separate PR and then we will rebase yours.

KKopyscinski commented on tests/bluetooth/blob_io_flash/src/main.c:
```diff
@@ -0,0 +1,256 @@
+	err = blob_flash_stream.io.rd(&blob_flash_stream.io, &xfer, &block, &chunk);
+	zassert_false(err == 0, "Read outside flash area successful");
+
+	blob_flash_stream.io.close(&blob_flash_stream.io, &xfer);
```
I wanted to verify something here, but for now this return's `void` and `` is noop. I added this so it won't break in the future - maybe it'll not be noop at some point - so it'll at least "check" for crash

KKopyscinski commented:
> Looks very good. One suggestion. I don't think it is good to litter in `test/bluetooth` folder because we can start adding more unit tests. So I'd suggest to move the app located in `test/bluetooth/mesh` to `tests/bluetooth/mesh/basic` (I couldn't give any better name, welcome with suggestions) and then move the new test to `test/bluetooth/mesh/blob_io_flash`? Or I can do this in a separate PR and then we will rebase yours.

I didn't like locating this in `tests/blutooth` very much, but couldn't find better place for this. Looking ath `mesh` test app I agree that it could be moved to `basic`. I think I wanted to add this into `tests/bluetooth/mesh` but it was occupied :) I'll move folders and update commit

KKopyscinski commented:
> Looks very good. One suggestion. I don't think it is good to litter in `test/bluetooth` folder because we can start adding more unit tests. So I'd suggest to move the app located in `test/bluetooth/mesh` to `tests/bluetooth/mesh/basic` (I couldn't give any better name, welcome with suggestions) and then move the new test to `test/bluetooth/mesh/blob_io_flash`? Or I can do this in a separate PR and then we will rebase yours.

I didn't like locating this in `tests/blutooth` very much, but couldn't find better place for this. Looking ath `mesh` test app I agree that it could be moved to `basic`. I think I wanted to add this into `tests/bluetooth/mesh` but it was occupied :) I'll move folders and update commit

KKopyscinski committed:
Tests: Bluetooth: Make space for Mesh tests

This commit moves app from /tests/bluetooth/mesh to
tests/bluetooth/mesh/basic. This allows to reuse /tests/bluetooth/mesh
directory as location for more tests.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

KKopyscinski commented on tests/bluetooth/blob_io_flash/src/main.c:
```diff
@@ -0,0 +1,256 @@
+	off_t block_idx = 0;
+	uint16_t chunk_idx = 0;
+	uint8_t chunk_data[BLOB_CHUNK_SIZE_MAX(BT_MESH_TX_SDU_MAX)];
+	uint8_t test_data[SLOT1_PARTITION_SIZE];
```
As discussed offline, this relates to chunk length being not aligned with word length in flash. I added `#define CHUNK_SIZE` value and comment describing it so it won't get changed in the future without reason. With size of 65 and word length of 4 first chunk will have 3 bytes of padding at the end, and next chunks will have 1 byte off padding at start and 2 on end. Reads after writes assure that data was written correctly, despite padding 0xFF's overlap previously written data.

PavelVPV replied:
One last thing. Can we check that nothing is written at the padding?

KKopyscinski replied:
I added this check. Also, I saw that I haven't used defined `CHUNK_SIZE` everywhere that's applicable, so made that change.

alxelax commented on tests/bluetooth/blob_io_flash/src/main.c:
```diff
@@ -0,0 +1,256 @@
+	err = bt_mesh_blob_io_flash_init(&blob_flash_stream,
+					 SLOT1_PARTITION_ID, 0);
+	zassert_equal(err, 0, "BLOB I/O init failed with err=%d", err);
+	flash_area_close(fa);
```
Logically, `flash_area_close(fa);` should be called before `bt_mesh_blob_io_flash_init` since `flash_area_..` functions perform access to flash map. The access should be logically completed. Of course, right now, `flash_area_close` does nothing and really doesn't matter to call it at all. However, if somebody in the future will add any logic to this function(at least function header description assumes that), the test behavior might be broken.

I'd suggest to add `flash_area_close(fa);` before `bt_mesh_blob_io_flash_init` to prevent issues with maintenance.

KKopyscinski replied:
makes sense to me, moved this call as you suggested

alxelax reviewed (commented):
LGTM. Valgrind doesn't show any memory issues (apart Kernel bugs :))

Please add `flash_area_close` in correct place if @PavelVPV doesn't have disagreement with it. I'll approve.
